### ISSUE TITLE
Plugin development: Edit workspace content

### ISSUE THREAD

**Issue by lunix33** [getinsomnia/insomnia#2983]
Hi, I'm currently in the process of creating a small plugin.

Currently all my plugin actions are bound to the workspace, but I was wondering if it was possible to edit its content.
By example I currently have an action able to import an API into the current workspace though OpenAPI definition.

But what if I want to change some of the requests properties, cookies or environment variables once the import is done.

The workspace context doesn't seems to allow those kind of action, and the models seems to be plain object and changes to said object doesn't affect the actual element. (see following example)

```js
async (context, {requests}) => {
	const myReq = requests.find((r) => r.name === 'My Request');

	myReq.headers.length = 0;

	// ...
	// Once the function is done, the request sill has a header attached.
}
```

Even if I'd be able to make change using the models I can't seem to be able to edit the cookies nor the environment variable.

Is there a way to go around that? have I missed some important documentation? Is it planned for the future?

**Comment by apisaint** [comment]
I've read through your comment and this point particular stands out to me:

> Even if I'd be able to make change using the models I can't seem to be able to edit the cookies nor the environment variable.

However in your code sample you do the following:

`myReq.headers.length = 0;`

This wouldn't change the headers in the way that you're expecting. You seem to want to empty the headers out; I haven't looked at the type of the `headers` property but you're probably looking at doing something like:

```js
myReq.headers = []; // for an array
// myReq.headers = {}; // when an object
```

Another thing, given that you can already import using OpenAPI, what is missing or are you looking to do with the plugin so those responding could better assist? And have you looked into `insomnia-importers` package?

**Comment by lunix33** [comment]
I just showed clearing the `headers` array as an example, but this applies to any properties of the model.
Any change to it doesn't change the displayed request in the UI.

The plugin is made to handle authentication and import (using OpenAPI definition) for an app internal API.
So far like I said I'm able to do import the API without any issue with the `context.data.import.uri()`, but the imported API actually has a few error (Could be another issue, since the endpoints `security` property doesn't seems to be properly respected), some of the endpoints which doesn't require authentication headers actually have them, so I'd like to remove them. Those changes I'd like them to be displayed in the UI. The issue is that I can't figure out if there's a way to actually save changes I made to the models.

And I also need to be able to change environment variable and cookies to have the authentication working properly, I want the users to have as little manipulation as possible to do to get the authentication working. But for those two it seems like there's no model available at the workspace level to read the values and like for the request, once the changes are done I need a way to save them so the UI would be able to see the new values.

Regards.

**Comment by apisaint** [comment]
I see. I think the best approach would be to create a PR for the [`insomnia-importers`](https://github.com/Kong/insomnia/tree/develop/packages/insomnia-importers) package, since this is what creates the requests in the first place, maybe there is a bug.

That said, I'm not well versed in the plugin API yet, but a quick look through [`insomnia-plugin-default-headers`](https://github.com/Kong/insomnia/blob/develop/plugins/insomnia-plugin-default-headers/src/default-headers.js) shows the usage of the following API to modify a request:

```js
    context.request.setHeader(name, value);
    console.log(`[header] Set default header ${name}: ${value}`);
```

Perhaps the `request` item in the `requests` array also has a `setHeader` function.

**Comment by Ghosty-chan** [comment]
Hi, I'm having similar issues, and was wondering if there's a solution except running it as a "Requesthook" similar like the default-headers are doing it..

**Background:**
As I've "many" requests with Auth: Bearer, I wanted to switch all of them to OAuth 2. (& Later to export them 'correctly')
Now I gotta just add those 5-7 Variables everytime - not so hard when doing a new request.. but replacing those existing ones.. or after importing some Postman Collections - I gotta do all of them again.. 
Since there's no simple "Copy & Paste" of those whole tabs, nor the bulk-header editor working in this case
I've started doing a plugin and..

**My Current State**
- Single Request (using **requestActions**): 

1. Copy
2. Replace another Request's Auth
3. Re-opening/Refreshing that Request (as React isn't picking up those changes.. fine for me)
Worked.

- Folder (using **requestGroupActions**):
1. Copy from one request (works)
2. Replace multiple Requests Auths - console.log proved me that it "should have worked"
4. Reopen/Refresh that request or any other one inside that folder -> still old Auth "in-use"
5. Restart Insomnia -> also shows nothing changed

I'm assuming that I'm just altering some "shadow" copies of those requests while using the requestGroupActions, while not beeing able to "save" those back to the app?

**(not working) Code** 
```js
module.exports.requestGroupActions = [
  { label: 'Replace all authentications',
    icon: 'fa-exchange',
    action: async (context, {requests}) => {
      let authDetails = await context.store.getItem('authDetails');
      for (const request of requests) {
        request.authentication = authDetailsObject;
      }
  }
}
...
]
```

**Single Request Code** 
```js
module.exports.requestActions = [
{   label: 'Quick Replace: Auth',
    icon: 'fa-exchange',
    action: async (context, data) => {
      const { request } = data; //authentication
      let authDetails = await context.store.getItem('authDetails');
      request.authentication = JSON.parse(authDetails);
    }
}
...
]
```
